src/lib/teacher-network.ts

```diff
@@ -0,0 +1,134 @@
+  if (!promise) {
+    promise = classTeachersPromises[classUri] = new Promise((resolve, reject) => {
+      const fetchOptions: RequestInit = { headers: { Authorization: `Bearer/JWT ${rawPortalJWT}` } };
+      fetch(classUri, fetchOptions)
+        .then(response => {
+          if (response.ok) {
+            response.json().then((portalClass: IPortalClassInfo) => {
+              resolve(portalClass.teachers.map(t => `${t.user_id}`));
+            });
+          }
+          else {
+            resolve(undefined);
+          }
+        })
+        .catch(error => {
+          resolve(undefined);
+        });
+    });
+  }
+  return promise;
+}
+
+// converts "msa", "1.4" to "msa/1/4"
+export function getProblemPath(unit: string, problem: string) {
+  return [unit, ...problem.split(".")].join("/");
+}
+
+// synchronize the current teacher's classes and offerings to firestore
+export function syncTeacherClassesAndOfferings(firestore: Firestore, user: UserModelType, rawPortalJWT: string) {
+  const { network } = user;
+  if (!network) return [];
+
+  const promises: Promise<void>[] = [];
+
+  // map portal offerings to classes
+  const userClasses: Record<string, ClassWithoutTeachers> = {};
+  user.portalClassOfferings.forEach(offering => {
+    const { classId: id, classHash: context_id, className: name, classUrl: uri, teacher } = offering;
+    if (!userClasses[context_id]) {
+      userClasses[context_id] = { id, context_id, name, uri, teacher, network };
+    }
+  });
+
+  // synchronize the classes
+  Object.keys(userClasses).forEach(async classHash => {
+    promises.push(syncClass(firestore, rawPortalJWT, userClasses[classHash]));
+  });
+
+  // synchronize the offerings
+  user.portalClassOfferings.forEach(async offering => {
+    const {
+      offeringId: id, activityTitle: name, activityUrl: uri, classHash: context_id, classUrl,
+      unitCode: unit, problemOrdinal: problem
+    } = offering;
+    const problemPath = getProblemPath(unit, problem);
+    const fsOffering: OfferingWithoutTeachers = { id, name, uri, context_id, unit, problem, problemPath, network };
+    promises.push(syncOffering(firestore, rawPortalJWT, classUrl, fsOffering));
+  });
+
+  return promises;
+}
+
+export async function syncClass(firestore: Firestore, rawPortalJWT: string, aClass: ClassWithoutTeachers) {
+  const { uri, context_id, network } = aClass;
+  if (uri && context_id && network && rawPortalJWT) {
+    const classDoc = firestore.doc(`classes/${network}_${context_id}`);
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const fsClass = await classDoc.get();
+      // class already exists in firestore; could sync contents but we'll skip for now
+    }
+    catch(e) {
+      if (isFirestorePermissionsError(e)) {
+        const teachers = await getClassTeachers(uri, rawPortalJWT);
+        if (teachers) {
+          aClass.teachers = teachers;
+          classDoc.set(aClass);
```

Comment:
This is approach is a bit hard to follow. It might also result in errors in the firestore logs which will be confusing when triaging issues. I don't understand all of this well enough, but perhaps the rules can be updated so that the `get()` will not throw a permissions error. And then you can use `fsClass.exists()` to see if the document already exists. 

Alternatively you could implement the sync which would then remove the line 
`const fsClass = await classDoc.get()` and then exception
 
So simply doing `classDoc.set(aClass)` will create or update the document hopefully without any errors thrown. If there is existing stuff in the class doc that isn't in aClass then `merge: true` can be used to not overwrite it.

Reply:
>It might also result in errors in the firestore logs

I checked that it doesn't result in errors in the browser's debugger console, so I wouldn't expect it to result in errors in the firestore logs, but I should check that as well.

>perhaps the rules can be updated so that the `get()` will not throw a permissions error

I don't think that's possible. It is inherent in this approach that we restrict access to documents based on the contents of those documents, and in that situation security rules throw a permissions error when requesting a nonexistent document. In a previous iteration of the code I used an `onSnapshot()` handler and its separate success and error callbacks but that seemed even more confusing.

>So simply doing classDoc.set(aClass) will create or update the document hopefully without any errors thrown.

It seems wrong to me to update/rewrite a bunch of database entries on every launch of the application even though we expect that information will rarely (if ever) be changed. Furthermore, acknowledging that this should not be an overriding concern at this point, writes are priced at three times the cost of reads, so reading first to decide whether a write is necessary is more cost-effective for documents that don't change often as well.

>I guess a 3rd option here is to refactor this into a function with a meaningful name that can be used in both places like getOrCreateDoc

Encapsulating this into a function seems like a reasonable compromise. I'll look into that as well as checking what the firestore logs reveal.

Reply:
I'm not sure it is worth it, or whether it would work in practice, but in theory there could be a rule that allows read if the document doesn't exist. This should mean that `doc(path).get()` when there is no document at the path, would no longer throw a security error. And since reading non-existing documents doesn't expose any data it seems like a safe thing to do. The downside I see is more complexity in the security rules, and perhaps the cost or performance hit of an extra read.

Someone asking about this is here: https://stackoverflow.com/questions/49876921/allow-read-of-firestore-document-if-it-does-not-exist they don't explain what they are doing clearly though, and they never confirm if the suggested syntax fix actually worked.

Reply:
I added a commit that refactors the common code into a `guaranteeDocument()` function that encapsulates the logic.

>I'm not sure it is worth it, or whether it would work in practice, but in theory there could be a rule that allows read if the document doesn't exist.

That's an interesting approach and it _might_ work, but given a choice between handling the additional complexity client-side (e.g. the `guaranteeDocument()` approach referenced above) or with more complex security rules I think I'd opt for simpler security rules.